### PR TITLE
pythonPackages.matplotlib: fix headless detection

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change

The default backend is chosen based on the content of the $DISPLAY
variable *and* a successfull call to libX11, loaded via dlopen().
The test fails because dlopen looks in /usr/lib and /lib, so matplotlib
falls back to a headless backend.

To reproduce try running:

    $ nix-shell -I nixpkgs=$PWD -p \
      'python3.withPackages (p: [ p.matplotlib ])' --run python
    >>> import matplotlib.pyplot as plt
    >>> assert plt.get_backend() == "TkAgg"

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via the above method
- [ ] Tested compilation of all pkgs that depend on this change
- [x] Tested execution of all binary files (none)
- [x] Added a release notes entry if the change is major or breaking (it's not)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 124439` at fa2fa9db run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>57 packages marked as broken and skipped:</summary>
  <ul>
    <li>cassandra_2_1</li>
    <li>cassandra_2_2</li>
    <li>hadoop</li>
    <li>hadoop_2_7</li>
    <li>hadoop_2_8</li>
    <li>hadoop_2_9</li>
    <li>hydrus</li>
    <li>impressive</li>
    <li>linuxPackages-libre.virtualbox</li>
    <li>linuxPackages_hardkernel_4_14.virtualbox</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/06d68612bde451584941e5ea0434b5c9">python38Packages.kmapper</a></li>
    <li><a href="https://gist.github.com/61fd3bc49328adbfaa4d98f63c91f537">python38Packages.python-igraph</a></li>
    <li><a href="https://gist.github.com/2437e364c5e1d8d32cf234f7ec99e97f">python38Packages.scikit-tda</a></li>
  </ul>
</details>
<details>
  <summary>363 packages skipped due to time constraints:</summary>
  <ul>
    <li>DisnixWebService</li>
    <li>R</li>
    <li>abcl</li>
    <li>adapta-gtk-theme</li>
    <li>adoptopenjdk-icedtea-web</li>
    <li>alloy (alloy5)</li>
    <li>alloy4</li>
    <li>antlr</li>
    <li>antlr3 (antlr3_5)</li>
    <li>antlr3_4</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>164 packages built successfully:</summary>
  <ul>
    <li>anki</li>
    <li>ankisyncd</li>
    <li>asciidoc-full-with-plugins</li>
    <li>corrscope</li>
    <li>datasette (python38Packages.datasette)</li>
    <li>deeptools</li>
    <li>evillimiter</li>
    <li>fdroidserver</li>
    <li>flatcam</li>
    <li>fop</li>
    <li>graph-cli</li>
    <li>igraph</li>
    <li>image-analyzer</li>
    <li>jre8</li>
    <li>liberation-sans-narrow</li>
    <li>liberation_ttf (liberation_ttf_v2)</li>
    <li>mavproxy</li>
    <li>mnemosyne</li>
    <li>pitivi</li>
    <li>psrecord</li>
    <li>pyspread</li>
    <li>python38Packages.Pweave</li>
    <li>python38Packages.androguard</li>
    <li>python38Packages.ase</li>
    <li>python38Packages.basemap</li>
    <li>python38Packages.bayespy</li>
    <li>python38Packages.booleanoperations</li>
    <li>python38Packages.castepxbin</li>
    <li>python38Packages.colorcet</li>
    <li>python38Packages.dask-image</li>
    <li>python38Packages.deap</li>
    <li>python38Packages.defcon</li>
    <li>python38Packages.dufte</li>
    <li>python38Packages.fipy</li>
    <li>python38Packages.fontmath</li>
    <li>python38Packages.fontparts</li>
    <li>python38Packages.fontpens</li>
    <li>python38Packages.fonttools</li>
    <li>python38Packages.glymur</li>
    <li>python38Packages.holoviews</li>
    <li>python38Packages.hvplot</li>
    <li>python38Packages.image-match</li>
    <li>python38Packages.imagecorruptions</li>
    <li>python38Packages.ipydatawidgets</li>
    <li>python38Packages.ipympl</li>
    <li>python38Packages.librosa</li>
    <li>python38Packages.lightning</li>
    <li>python38Packages.markdown-it-py</li>
    <li>python38Packages.matplotlib</li>
    <li>python38Packages.mdit-py-plugins</li>
    <li>python38Packages.memory_profiler</li>
    <li>python38Packages.mplfinance</li>
    <li>python38Packages.mutatormath</li>
    <li>python38Packages.nbsmoke</li>
    <li>python38Packages.nbval</li>
    <li>python38Packages.osmpythontools</li>
    <li>python38Packages.persim</li>
    <li>python38Packages.phik</li>
    <li>python38Packages.pims</li>
    <li>python38Packages.pint</li>
    <li>python38Packages.psautohint</li>
    <li>python38Packages.pyalgotrade</li>
    <li>python38Packages.pycm</li>
    <li>python38Packages.pycocotools</li>
    <li>python38Packages.pymatgen</li>
    <li>python38Packages.pyregion</li>
    <li>python38Packages.pyspice</li>
    <li>python38Packages.pytest-astropy</li>
    <li>python38Packages.pytest-astropy-header</li>
    <li>python38Packages.pytest-mpl</li>
    <li>python38Packages.pytest-regressions</li>
    <li>python38Packages.python-csxcad</li>
    <li>python38Packages.python-openems</li>
    <li>python38Packages.pytomlpp</li>
    <li>python38Packages.qcelemental</li>
    <li>python38Packages.regional</li>
    <li>python38Packages.ripser</li>
    <li>python38Packages.scapy</li>
    <li>python38Packages.scikit-bio</li>
    <li>python38Packages.scikitimage</li>
    <li>python38Packages.seaborn</li>
    <li>python38Packages.showit</li>
    <li>python38Packages.skidl</li>
    <li>python38Packages.slicedimage</li>
    <li>python38Packages.spyder-kernels_0_5</li>
    <li>python38Packages.spyder_3</li>
    <li>python38Packages.stytra</li>
    <li>python38Packages.tadasets</li>
    <li>python38Packages.ufoprocessor</li>
    <li>python38Packages.vidstab</li>
    <li>python38Packages.wordcloud</li>
    <li>python39Packages.Pweave</li>
    <li>python39Packages.androguard</li>
    <li>python39Packages.ase</li>
    <li>python39Packages.bayespy</li>
    <li>python39Packages.booleanoperations</li>
    <li>python39Packages.castepxbin</li>
    <li>python39Packages.colorcet</li>
    <li>python39Packages.dask-image</li>
    <li>python39Packages.datasette</li>
    <li>python39Packages.deap</li>
    <li>python39Packages.defcon</li>
    <li>python39Packages.dufte</li>
    <li>python39Packages.fontmath</li>
    <li>python39Packages.fontparts</li>
    <li>python39Packages.fontpens</li>
    <li>python39Packages.fonttools</li>
    <li>python39Packages.glymur</li>
    <li>python39Packages.holoviews</li>
    <li>python39Packages.hvplot</li>
    <li>python39Packages.image-match</li>
    <li>python39Packages.imagecorruptions</li>
    <li>python39Packages.ipydatawidgets</li>
    <li>python39Packages.ipympl</li>
    <li>python39Packages.kmapper</li>
    <li>python39Packages.lightning</li>
    <li>python39Packages.markdown-it-py</li>
    <li>python39Packages.matplotlib</li>
    <li>python39Packages.mdit-py-plugins</li>
    <li>python39Packages.memory_profiler</li>
    <li>python39Packages.mplfinance</li>
    <li>python39Packages.mutatormath</li>
    <li>python39Packages.nbsmoke</li>
    <li>python39Packages.nbval</li>
    <li>python39Packages.osmpythontools</li>
    <li>python39Packages.perfplot</li>
    <li>python39Packages.persim</li>
    <li>python39Packages.pims</li>
    <li>python39Packages.pint</li>
    <li>python39Packages.psautohint</li>
    <li>python39Packages.pyalgotrade</li>
    <li>python39Packages.pycm</li>
    <li>python39Packages.pycocotools</li>
    <li>python39Packages.pymatgen</li>
    <li>python39Packages.pyregion</li>
    <li>python39Packages.pyspice</li>
    <li>python39Packages.pytest-astropy</li>
    <li>python39Packages.pytest-astropy-header</li>
    <li>python39Packages.pytest-mpl</li>
    <li>python39Packages.pytest-regressions</li>
    <li>python39Packages.python-igraph</li>
    <li>python39Packages.pytomlpp</li>
    <li>python39Packages.qcelemental</li>
    <li>python39Packages.regional</li>
    <li>python39Packages.ripser</li>
    <li>python39Packages.scapy</li>
    <li>python39Packages.scikit-bio</li>
    <li>python39Packages.scikit-fuzzy</li>
    <li>python39Packages.scikitimage</li>
    <li>python39Packages.seaborn</li>
    <li>python39Packages.showit</li>
    <li>python39Packages.skidl</li>
    <li>python39Packages.slicedimage</li>
    <li>python39Packages.spyder-kernels_0_5</li>
    <li>python39Packages.spyder_3</li>
    <li>python39Packages.tadasets</li>
    <li>python39Packages.ufoprocessor</li>
    <li>python39Packages.vidstab</li>
    <li>python39Packages.wordcloud</li>
    <li>python39Packages.xgboost</li>
    <li>scfbuild</li>
    <li>tartube</li>
    <li>visidata</li>
    <li>xits-math</li>
  </ul>
</details>
<details>
  <summary>6 suggestions:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `nose`.
Near pkgs/development/python-modules/matplotlib/default.nix:2:28:
```
  |
2 | , which, cycler, dateutil, nose, numpy, pyparsing, sphinx, tornado, kiwisolver
  |                            ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/license-missing.md">license-missing</a>

Package is missing a license.

Near pkgs/development/python-modules/matplotlib/default.nix:81:3:
```
   |
81 |   meta = with lib; {
   |   ^
```
</li>
    <li>warning: unused-argument

Unused argument: `pythonOlder`.
Near pkgs/development/python-modules/matplotlib/default.nix:13:3:
```
   |
13 | , pythonOlder
   |   ^
```
</li>
    <li>warning: unused-argument

Unused argument: `python`.
Near pkgs/development/python-modules/matplotlib/default.nix:1:38:
```
  |
1 | { lib, stdenv, fetchPypi, writeText, python, buildPythonPackage, isPy3k, pycairo, backports_functools_lru_cache
  |                                      ^
```
</li>
    <li>warning: unused-argument

Unused argument: `backports_functools_lru_cache`.
Near pkgs/development/python-modules/matplotlib/default.nix:1:83:
```
  |
1 | { lib, stdenv, fetchPypi, writeText, python, buildPythonPackage, isPy3k, pycairo, backports_functools_lru_cache
  |                                                                                   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md">python-include-tests</a>

Consider adding a `checkPhase` for tests, or if not feasible, `pythonImportsCheck`.

Near pkgs/development/python-modules/matplotlib/default.nix:82:0:
```
   |
82 |     description = "Python plotting library, making publication quality plots";
   | ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

File: pkgs/development/python-modules/matplotlib/default.nix [hunk 1]
```diff
@@ -62,8 +64,15 @@ buildPythonPackage rec {
     let
       tcl_tk_cache = ''"${tk}/lib", "${tcl}/lib", "${lib.strings.substring 0 3 tk.version}"'';
     in
-    lib.optionalString enableTk
-      "sed -i '/self.tcl_tk_cache = None/s|None|${tcl_tk_cache}|' setupext.py";
+    lib.optionalString enableTk ''
+      sed -i '/self.tcl_tk_cache = None/s|None|${tcl_tk_cache}|' setupext.py
+    '' +
+    lib.optionalString stdenv.isLinux ''
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/matplotlib/default.nix:70]:
```suggestion
    '' + lib.optionalString stdenv.isLinux ''
```

**Comment by SuperSandro2000** [comment]
We probably should rebase this against staging.

**Comment by CMCDragonkai** [comment]
This just bit me. The release 21.05 has this problem. It should be backported to release 21.05.

File: pkgs/development/python-modules/matplotlib/default.nix (hunk 1 above)
Comment (veprbl) [inline pkgs/development/python-modules/matplotlib/default.nix:70]: Shoud this be `stdenv.isLinux && (enableGtk3 || enableTk || enableQt)`?

Comment (rnhmjoj) [reply pkgs/development/python-modules/matplotlib/default.nix:70]: Right.

**Review by veprbl — APPROVED** [review]
LGTM

**Comment by veprbl** [comment]
@GrahamcOfBorg build python3Packages.matplotlib

**Comment by CMCDragonkai** [comment]
How to port this to 21.05?

**Comment by veprbl** [comment]
> How to port this to 21.05? #122042

Already done in #124488

**Pull request merged**